### PR TITLE
Budget Manager

### PR DESCRIPTION
**Pull request by sagieramos**
#### Design
- You should follow these [design guidelines](https://www.behance.net/gallery/19759151/Snapscan-iOs-design-and-branding?tracking_source=), including:
  - Colors.
  - Typography: font face, size and weight.
  - Layout: composition and space between elements.

> NOTE: In these design guidelines there are several UIs that you won't need for this exercise; also, some pages are not given a design and you will create them following the design guidelines of the other pages.

Original design idea by [Gregoire Vella on Behance](https://www.behance.net/gregoirevella).

The [Creative Commons license of the design](https://creativecommons.org/licenses/by-nc/4.0/) requires that you give appropriate credit to the author. Therefore, you must do it in the README of your project.

#### Interactions
- Splash screen
  - A simple page with the name of your app (yes, you need to choose one), and links to the sign up and log in pages.

- Sign up and log in pages
  - The user should be able to register in the app with full name, email and password (all mandatory).
  - The user can log into the app using email and password.
  - If the user is not logged in, they can't access pages that require the user to be logged in (all the pages described below).

- Home page (categories page)
  - When the user logs in, they are presented with the categories page.
  - For each category, the user can see their name, icon and the total amount of all the transactions that belongs to that category.
  - When the user clicks (or taps) on a category item, the application navigates to the transactions page for that category.
  - There is a button "add a new category" at the bottom that brings the user to the page to create a new category.

- Transactions page
  - For a given category, the list of transactions is presented, ordered by the most recent.
  - At the top of the page the user could see the total amount for the category (sum of all of the amounts of the transactions in that category).
  - There is a button "add a new transaction" at the bottom that brings the user to the page to create a new transaction.
  - When the user clicks on the "Back" button (<), the user navigates to the home page.

- "Add a new category" page
  - The user fills out a form to create a new category, indicating their name and icon (both mandatory).
  - The user clicks (or taps) the "Save" button to create the new category, and is taken to the home page on success.
  - When the user clicks on the "Back" button (<), the user navigates to the home page.

- "Add a new transaction" page
  - The user fills out a form to create a new transaction with:
    - name (mandatory)
    - amount (mandatory)
    - categories (mandatory at least one)
  - The user click (or taps) the "Save" button to create the new transaction, and is taken to the transactions page for that category.
  - When the user clicks on the "Back" button (<), the user navigates to the transactions page for that category.

#### Testing requirements
- Create unit and integration tests for all the most important components of your RoR application.

#### Technical requirements

- You should use Postgres as your database.
- You should use devise for authentication.
- You should validate all user input to make sure that anyone with bad intentions cannot compromise your app.
- You can use a view template engine of your choice (.erb, .slim, .haml).
- The project should be deployed and accessible online.
- Your database schema should reflect the following structure:

### CODE REVIEW

File: app/views/entities/show.html.erb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+  <div>
+      <% if @entity.groups.any? %>
+          <h3 class="h3-title"> Categories: </h3>
+          <%= @entity.groups.map { |group| link_to("\##{group.name}", group_path(group), class: "group-name-sm") }.join(' | ').html_safe %>
+      <% end %> 
+  </div>
+
+  <hr>
+
```
Comment (Munsa1) [inline app/views/entities/show.html.erb:27]: - [ ] Great job so far, however kindly follow the given design guidelines, in terms of:

- Colors
- Typography: font face, size, and weight.
- Layout: composition and space between elements.

The colors you have used are close but please follow the given color swatches and layout of elements [here](https://www.behance.net/gallery/19759151/Snapscan-iOs-design-and-branding?tracking_source=&). For example the `add category` and add `transaction` button is aligned at the bottom


Given design
image

Your design
image

Also the given font family is Proxima Nova, kindly use it for your entire design

File: app/views/groups/show.html.erb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    <table>
+      <thead>
+        <tr>
+          <th>Transactions (<%= @group.entities.count %>)</th>
+          <th>Amount</th>
+          <th>Date</th>
+        </tr>
+      </thead>
+      <tbody>
+        <% @group.entities.each do |entity| %>
```
Comment (Munsa1) [inline app/views/groups/show.html.erb:21]: - [x] Transactions should be listed or presented in order by the most recent. Kindly do so. If you look at the screenshot below the time stamp for the second transaction from the top is more recent than the first one, yet it is second in the order.
image

File: app/views/groups/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  <br />
+  <div>
+    <% if @group.user == current_user %>
+      <%= link_to 'Add transaction', new_entity_path(group_id: @group.id) %> |
+      <%= link_to 'Edit', edit_group_path(@group) %> |
+      <%= link_to 'Delete', group_path(@group), method: :delete, data: { confirm: 'Are you sure you want to delete this group?' } %>
+    <% end %>
+  </div>
+
+  <%= link_to 'Back Categories', groups_path %>
```
Comment (Munsa1) [inline app/views/groups/show.html.erb:45]: - [ ] If the user goes to the form to create a transaction and clicks the back arrow, they should be taken to the home page and not the `all transactions` page. Kindly fix this.

Screen Cast
[cinnamon-2024-01-13T154757+0200.webm](https://github.com/sagieramos/BudgetManager/assets/81932478/ba90feec-174b-4f55-af6f-938623e72ab3)

File: app/views/groups/show.html.erb [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+            <td><%= entity.created_at.strftime('%Y-%m-%d %H:%M:%S') %></td>
+          </tr>
+        <% end %>
+      </tbody>
+    </table>
+    <br />
+    <h3 class="h3-title">Total Amount: <%= format_currency(@group.entities.sum(:amount)) %></h3>
+  <% else %>
+    <p>No transaction</p>
+  <% end %>
```
Comment (Munsa1) [inline app/views/groups/show.html.erb:34]: - [ ] When creating a transaction, a user should be presented with a form that they will have to fill out:

* name (mandatory)
* amount (mandatory)
* categories (mandatory at least one)

Currently, your approach is that the category is assigned automatically, as it is a requirement that the category also be selected manually by the user kindly implement this


screencast
[cinnamon-2024-01-13T155359+0200.webm](https://github.com/sagieramos/BudgetManager/assets/81932478/c8830fd0-ee7f-4d9c-a64f-e2e2fc23a542)

**Review by Munsa1 — CHANGES_REQUESTED** [review]
Hi @sagieramos, great video presentation
required changes

Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation but you are almost there!
<FEEL FREE TO CHANGE THIS MESSAGE TO YOUR PERSONAL WORDS>


## Suggested changes
_Check the comments under the review._

You can use as many of my suggestions as you want. If there is anything you would like to skip - feel free to do that. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

File: app/views/groups/show.html.erb (hunk 4 above)
Comment (sagieramos) [reply app/views/groups/show.html.erb:34]: When creating a new transaction from a current category, the newly created transaction is automatically added to that category. 

To explicitly select categories for the newly created transaction, click on the hamburger icon, then `add transaction`, and you will be provided with a form:

Screenshot from 2024-01-13 15-11-11

Screenshot from 2024-01-13 15-11-40

when you edit a transaction, you will receive the same form that enables you to select more categories. You will see the categories that it initially belongs to already selected. You can uncheck it and the transaction will automatically disassociate from that category when you submit

File: app/views/groups/show.html.erb (hunk 3 above)
Comment (sagieramos) [reply app/views/groups/show.html.erb:45]: This was resolved here:

Screenshot from 2024-01-13 15-39-07

I have removed lines 11 and 12

File: app/views/groups/show.html.erb (hunk 2 above)
Comment (sagieramos) [reply app/views/groups/show.html.erb:21]: Added line 13:

Screenshot from 2024-01-13 15-58-57

changes line 21:

Screenshot from 2024-01-13 16-00-39

**Review by MahmoudBakr23 — APPROVED** [review]
Hi @sagieramos 😃, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit:
Congratulations! 🎉

**Highlights**
- Well done on resolving the issues mentioned by the previous reviewer
- Excellent communication

Cheers. and Happy coding!👏👏👏

Feel free to mention @MahmoudBakr23 to leave any questions or comments in the PR thread if something is not 100% clear.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**